integrations/slack/slack.go

```diff
@@ -0,0 +1,126 @@
+
+// request represents a typical HTTP request sent to the Slack API.
+type request struct {
+	Channel     string      `json:"channel"`
+	Text        string      `json:"text"`
+	Attachments interface{} `json:"attachments"`
+}
+
+// Attachment represents a basic Slack Attachment
+type Attachment struct {
+	Fallback   string             `json:"fallback"`
+	Color      string             `json:"color"`
+	Pretext    string             `json:"pretext"`
+	AuthorName string             `json:"author_name"`
+	AuthorLink string             `json:"author_link"`
+	AuthorIcon string             `json:"author_icon"`
+	Title      string             `json:"title"`
+	TitleLink  string             `json:"title_link"`
+	Text       string             `json:"text"`
+	Fields     []*AttachmentField `json:"fields"`
+	ImageUrl   string             `json:"image_url"`
+	ThumbUrl   string             `json:"thumb_url"`
+	Footer     string             `json:"footer"`
+	FooterIcon string             `json:"footer_icon"`
+	Ts         int64              `json:"ts"`
+}
+
+// AttachmentField represents Slack fields.
+type AttachmentField struct {
+	Title string `json:"title"`
+	Value string `json:"value"`
+	Short bool   `json:"short"`
+}
+
+// response represents the returned HTTP response.
+type response struct {
+	Ok    bool   `json:"ok"`
+	Error string `json:"error"`
+}
+
+// Config stores client configuration.
+type Config struct {
+	URI   string
+	Token string
+}
+
+// Client contains a http.Client along with config.
+type Client struct {
+	*http.Client
+	config *Config
+}
+
+// NewClient returns a Client.
+func NewClient(token string) *Client {
+	return &Client{
+		Client: http.DefaultClient,
+		config: &Config{
+			URI:   URI,
+			Token: token,
+		},
+	}
+}
+
+// Sends message to the specified channel.
+func (client *Client) SendMessage(ctx context.Context, channel, text string, attachments []*Attachment) error {
+	req := &request{
+		Channel:     channel,
+		Text:        text,
+		Attachments: attachments,
+	}
+
+	// marshal request into JSON
+	b, err := json.Marshal(req)
+	if err != nil {
+		return err
+	}
+
+	return client.sendResponse(ctx, b)
+}
+
+// Sends a HTTP request to the URI and decodes response.
+func (client *Client) sendResponse(ctx context.Context, b []byte) error {
+	// create new request
+	req, err := http.NewRequest(http.MethodPost, client.config.URI, bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+
+	// add headers
+	_ = req.WithContext(ctx)
+	req.Header.Set("Authorization", "Bearer "+client.config.Token)
+	req.Header.Set("Content-type", "application/json; charset=utf-8")
+
+	// send HTTP request
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	// decode response
+	var res response
+	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return err
+	}
+
+	// check response and return error
+	if !res.Ok {
+		return fmt.Errorf(res.Error)
+	}
+
+	return nil
```

Comment:
```suggestion
	return res
```
Probably return the response struct.

Reply:
I felt that returning errors would make more sense, as it could be used as such:

```go
	errCh := make(chan error, 1)

	go func() {
		errCh <- client.SendMessage(ctx, channel, text, nil)
	}()

	select {
	case err := <-errCh:
		if err != nil {
			return fail(err)
		}
		return 0
	case <-ctx.Done():
		return fail(ctx.Err())
	}
```

Reply:
Yup, but returning the response status is more expressive rather than returning `nil` when a request is made.

Reply:
Instead of returning `err`, is `log.Fatal(err)` the way to go?

Reply:
`log.Fatal(err)` is used for irrecoverable errors, or errors which have the capability to crash the program. Returning `log.Fatal(err)` doesn't make sense in this context. The user calling the method should take of whether to add `log.Fatal(err)`.

Reply:
Sorry for the misunderstanding, but I was asking your preference on how do you want an error (something like [this](https://github.com/deepsourcelabs/hermes/pull/12/files#diff-2eddbb7c9ab0d890aad5467d55547f9785773e12ce9db968095523a568a3e0daR109)) to be handled?

Do you want to log the error, etc.? Could you please specify the ideal behavior/way to handle errors in this context?

Reply:
You can return `err` when an error is encountered, otherwise return `res.Ok`.

Reply:
Thanks for clarifying the confusion! :)